### PR TITLE
Feature/17 provide logs from k8s loki

### LINKED ISSUE #17: Provide logs from k8s-loki
k8s-ces-control should provide logs from k8s-loki

### PR DESCRIPTION
**Pull request by kahoona77**
resolves #17

### CODE REVIEW

File: packages/logging/lokiLogProvider.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+	}
+	return defaultQueryLimit
+}
+
+// createQueryStartDateFromEndDate calculates the start date for a loki query based on the given end date.
+// Since loki query run backwards (in time) the start date is calculated based on the end date.
+// The start date is set 30 days before the given end date, because that is the maximum range that loki allows.
+func createQueryStartDateFromEndDate(endDate time.Time) time.Time {
+	return endDate.Add(-24 * 30 * time.Hour)
+}
```
Comment (jelemux) [inline packages/logging/lokiLogProvider.go:128]: Do we really want to query the maximum here? Cause that could be a lot of logs, most of which we'll never need.

File: packages/logging/lokiLogProvider.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+		}
+
+		// there are still logs to read -> start with the newest log timestamp from the last response
+		endDate = logLines[0].timestamp
+		startDate = createQueryStartDateFromEndDate(endDate)
+	}
+
+	// because multiple logs can happen at the exact same timestamp and the query is batched over time,
+	// it is possible that consecutive batches contain the same logLines. These need to be removed.
+	result = deduplicateLogLines(result)
```
Comment (jelemux) [inline packages/logging/lokiLogProvider.go:94]:
```suggestion
		endDate = logLines[0].timestamp.Add(-1 * time.Nanosecond)
		startDate = createQueryStartDateFromEndDate(endDate)
	}
```
Wouldn't this get rid of the need to deduplicate log lines?

**Review by jelemux — CHANGES_REQUESTED** [review]
Looks good but maybe we should talk about some details.

File: packages/logging/lokiLogProvider.go (hunk 1 above)
Comment (kahoona77) [reply packages/logging/lokiLogProvider.go:128]: The limit is always set to the linecount given by the user or to 1000 (when the given linecount is 0)
So there should never be more than 1000 lines in the result.

File: packages/logging/lokiLogProvider.go (hunk 2 above)
Comment (kahoona77) [reply packages/logging/lokiLogProvider.go:94]: let`s say that 5 logs happen at the exact same time. In the first batch the result only contains 3 of those 5 logs.
With this method we would loose the other 2 logs

File: packages/logging/lokiLogProvider.go (hunk 1 above)
Comment (jelemux) [reply packages/logging/lokiLogProvider.go:128]: I mean the line count in the request, which will be 30 days of logs if I'm understanding correctly. That will likely be a lot of network overhead.

Comment (jelemux) [reply packages/logging/lokiLogProvider.go:128]: Ah sorry, I understand now

File: Jenkinsfile [hunk 3]
```diff
@@ -94,6 +94,15 @@ node('docker') {
                 k3d.waitForDeploymentRollout("postfix", 300, 10)
             }
 
+            stage('Deploy k8s-ces-control dependencies') {
+                withCredentials([[$class: 'UsernamePasswordMultiBinding', credentialsId: 'harborhelmchartpush', usernameVariable: 'HARBOR_USERNAME', passwordVariable: 'HARBOR_PASSWORD']]) {
+                    k3d.helm("registry login ${registry} --username '${HARBOR_USERNAME}' --password '${HARBOR_PASSWORD}'")
+                    k3d.helm("install k8s-minio oci://${registry}/${registry_namespace}/k8s-minio --version 2023.9.23-3")
+                    k3d.helm("install k8s-loki oci://${registry}/${registry_namespace}/k8s-loki --version 2.9.1-3")
+                    k3d.helm("install k8s-promtail oci://${registry}/${registry_namespace}/k8s-promtail --version 2.9.1-2")
```
[the remark this answers is not in the document]
Comment (kahoona77) [reply Jenkinsfile:102]: Good idea. Will try :+1:

File: packages/logging/lokiLogProvider.go (hunk 2 above)
Comment (jelemux) [reply packages/logging/lokiLogProvider.go:94]: Yeah, I didn't understand that we request a limited number of logs from the server.

**Review by jelemux — APPROVED** [review]
LGTM!

**Pull request merged**